webpack: Enable code splitting and deduplication

andersk opened the pull request:
This sets up webpack [code splitting](https://webpack.js.org/guides/code-splitting/) using [SplitChunksPlugin](https://webpack.js.org/plugins/split-chunks-plugin/). We can now share sections of our code between different entry points—no more duplication between the `app` entry and the `common` entry, for example. Each page now uses a single webpack entry, split into multiple chunks—no more duplication between multiple entries included on one page. Best of all, webpack computes the split automatically, so we can go forward with modularization work without having to worry about this nonsense. The release tarball magically shrinks by 1.3 MB compressed, 4.9 MB uncompressed.

There is one caveat, hence the draft marker: code splitting necessarily entails giving up some control over the inclusion order of CSS files, which it turns out we were incorrectly relying on. Some of these bugs are known and not yet fixed here, mostly due to `media.scss`, whose rules must be included after the rules they override. There will probably be others that can only realistically be uncovered through QA.

**Testing Plan:** Dev server, production (https://andersk.zulipdev.org)

timabbott commented on tools/webpack.config.ts:
```diff
@@ -178,6 +179,11 @@ export default (env?: string): webpack.Configuration[] => {
+            splitChunks: {
+                chunks: "all",
+                maxAsyncRequests: 20,
+                maxInitialRequests: 20,
```
Can you add a comment on why these parameters should be 20?   I assume it should be 5 if you're not using HTTP 2, but we are?

andersk replied:
Right, these numbers came from [webpack/examples/many-pages/webpack.config.js](https://github.com/webpack/webpack/blob/v4.41.2/examples/many-pages/webpack.config.js) with `// for HTTP2` and seemed pretty reasonable. In practice it looks like we aren’t actually hitting the max so the exact numbers won’t matter as long as they’re large enough.

timabbott commented on tools/webpack.assets.json:
```diff
@@ -44,55 +69,98 @@
+        "./static/js/bundles/common.js",
+        "./static/js/translations.js",
+        "./static/js/portico/header.js",
+        "./static/styles/portico/portico-styles.scss",
```
Is there a way for this bundle of 4 lines to become just an include on the portico bundle?  Seems like they get duplicated a lot in this config.

andersk replied:
Ideally we’d move toward just one JS file per entry point that `import`s everything else it needs. I did this initial pass in the way that involved the least reasoning so I could automate it.

timabbott commented:
Overall approach looks great; posted a few small comments.  

For `media.scss`, one strategy is we could move a lot of other rules to SCSS inline media rules that sit next to the rule they're modifying (with whatever standard width constants we're using defined in a constants file).

andersk commented:
Addressed comments and known bugs. This should be approximately ready.

If we want to name the duplicated `@media` queries, we can use [postcss-custom-media](https://github.com/postcss/postcss-custom-media).

timabbott commented:
I merged the first half of this branch, with this change:
* Updated the commit message to avoid suggesting media CSS was ever in the file with the stuff it overrides except when we had a single CSS file with all the CSS.

For the 5xx.html commit, can you confirm you've tested that in a production environment?

andersk commented:
Yes, https://andersk.zulipdev.org/static/webpack-bundles/5xx.html

timabbott commented:
Great, merged the 5xx commit as well.  For `django-webpack-loader`, it looks like there's like a 0.0.5 https://pypi.org/project/django-webpack4-loader/; is there a reason you're using 0.0.1-beta?

timabbott commented:
Oh, oops, misread and was looking at the Node package.

andersk commented:
This uses django-webpack4-loader 0.0.5 (the Python part) with [webpack4-bundle-tracker](https://www.npmjs.com/package/webpack4-bundle-tracker) 0.0.1-beta (the JS part); both are current.

timabbott commented:
OK, looks great; I did some testing and merged this.  Will deploy to chat.zulip.org shortly so we can get complaints about anything the media changes broke.

We'll probably somewhat soon want to do a pass on `docs/subsystems/html-css.md` to update it to explain the new world order, but I'm fine with waiting on that until we've done a few more transitions.

Thanks @andersk!
